### PR TITLE
Add summary for GameGAN

### PR DESCRIPTION
**Pull request by praeclarumjj3**

### CODE REVIEW

**Comment by aniket-agarwal1999** [comment]
Hi,
So can u maybe look into the image explanations in your review, like the use of images should be a bit limited to only the essential ones(ideally 2-3).

**Comment by praeclarumjj3** [comment]
> Hi,
> So can u maybe look into the image explanations in your review, like the use of images should be a bit limited to only the essential ones(ideally 2-3).

Thanks for the reply. Are you counting the equations as images as well?

**Comment by praeclarumjj3** [comment]
@aniket-agarwal1999 I have made a new commit removing a few images. Please check again :)

File: summaries/GameGAN.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+GameGAN ingests screenplay and keyboard actions during training and aims to predict the next frame by conditioning on the action,i.e. a key pressed by the agent. It learns from rollouts of image and action pairs directly without having access to the underlying game logic or engine. GameGAN supports several applications such as transfer- ring a given game from one operating system to the other, without requiring to re-write code.
```
Comment (aniket-agarwal1999) [inline summaries/GameGAN.md:9]: *transferring

File: summaries/GameGAN.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+<img src='../images/GameGAN.png'>
+
+## GameGAN
+
+The focus is on an action-conditioned simulator in the image space where there is an egocentric agent that moves according to the given action *a<sub>t</sub> ∼ A* at time *t* and generates a new observation *x<sub>t+1</sub>*. We assume there is also a stochastic variable *z<sub>t</sub> ∼ N (0; I)* that corresponds to randomness in the environment. Given the history of images *x<sub>1:t</sub>* along with *a<sub>t</sub>* and *z<sub>t</sub>* , GameGAN predicts the next image *x<sub>t+1</sub>* . GameGAN is composed of three main modules: 
+
+###Dynamics Engine
```
Comment (aniket-agarwal1999) [inline summaries/GameGAN.md:17]: Space between ### and Dynamics, it is rendering as normal text because of no space.

File: summaries/GameGAN.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+- *h<sub>t</sub>*, *a<sub>t</sub>*, *z<sub>t</sub>*, *c<sub>t</sub>*, *x<sub>t</sub>* are the hidden state, action, stochastic variable, cell state, image at time step *t*. *m<sub>t-1</sub>* is the retrieved memory vector in the previous step (if memory module is used), and *i<sub>t</sub>*, *f<sub>t</sub>*, *o<sub>t</sub>* are the input, forget, and output gates. 
+
+- *a<sub>t</sub>*, *z<sub>t</sub>*, *m<sub>t-1</sub>* and *h<sub>t</sub>* are fused into *v<sub>t</sub>* , and *s<sub>t</sub>* is the encoding of the image *x<sub>t</sub>*. 
+
+- *H* is a MLP, *C* is a convolutional encoder, and *W* are weight matrices.
+
+- The engine maintains the standard state variables for LSTM, *h<sub>t</sub>* and *c<sub>t</sub>* , which contain information about every aspect of the current environment at time *t*. It computes the state variables given *a<sub>t</sub>*, *z<sub>t</sub>* , *m<sub>t</sub>* , and *x<sub>t</sub>*.
+
+###Memory Module
```
Comment (aniket-agarwal1999) [inline summaries/GameGAN.md:37]: Same as above here as well, space issue

File: summaries/GameGAN.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+- Simulating an environment in which there is an agent navigating through it, requires long-term consistency in which the simulated scene should not change when the agent comes back to the same location a few moments later. We propose to use an external memory module, motivated by the Neural Turing Machine (NTM).
+
+- The memory module has a memory block *M ∈ R<sup>N×N×D</sup>*, and the attended location *α<sub>t</sub> ∈ R<sup>N×N</sup>* at time *t*. *M* contains *N×N* *D-dimensional* vectors where *N* is the spatial width and height of the block. Intuitively, *α<sub>t</sub>* is the current location that the egocentric agent is located at. *M* is initialized with random noise ∼ *N (0, I)* and *α<sub>0</sub>* is initialized with 0s except for the **center location (N/2, N/2)** that is set to 1. At each time step, the memory module computes:
+
+<img src='../images/GameGAN_memory_eqn.png'>
+
+where; *K*, *G* and *E* are small MLPs. *w* is a learned shift kernel that depends on the current action, and the kernel is used to shift *α<sub>t−1</sub>*.
+
+###Rendering Engine
```
Comment (aniket-agarwal1999) [inline summaries/GameGAN.md:47]: Same here as well

File: summaries/GameGAN.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+- **Cycle Loss**
+
+    - To ensure long-term consistency of static elements, the memory module and the rendering engine are leveraged to disentangle static elements from dynamic elements.
+
+    - After running through some time steps *T* , the memory block *M* is populated with information from the dynamics engine. 
+    
+    - Using the memory location history *α<sub>t</sub>* , we can retrieve the memory vector *m̂<sub>t</sub>* which could be different from *m<sub>t</sub>* if the content at the location *α<sub>t</sub>* has been modified. 
+    
+    - Now, *c = { m̂ t , 0}* is passed to the rendering engine to produce *X<sup>m̂<sub>t</sub></sup>* where *0* is the zero vector and *X<sup>m̂<sub>t</sub></sup>* is the output component corresponding to *m̂<sub>t</sub>*. We use the following loss:
```
Comment (aniket-agarwal1999) [inline summaries/GameGAN.md:71]: Here *c = { m̂ t , 0}*, shouldn't it be *c = { m̂<sub>t</sub> , 0}*, I am not sure though. Please check once.

Comment (praeclarumjj3) [reply summaries/GameGAN.md:71]: Yes, you are right. Didn't add the <sub> tag there. Sorry!

File: summaries/GameGAN.md (hunk 4 above)
Comment (praeclarumjj3) [reply summaries/GameGAN.md:47]: Corrected... Thanks!

**Comment by aniket-agarwal1999** [comment]
Also @praeclarumjj3 is it paper from NVIDIA where they did the work on simulating PACMAN game? It was trending a few weeks back on twitter. The idea suggested in the paper seems pretty awesome :)

**Comment by praeclarumjj3** [comment]
> Also @praeclarumjj3 is it paper from NVIDIA where they did the work on simulating PACMAN game? It was trending a few weeks back on twitter. The idea suggested in the paper seems pretty awesome :)

Yes, it is the same one. 40 years of PACMAN!
The idea is really fascinating. These kinds of papers make DL a treat to pursue!

**Comment by aniket-agarwal1999** [comment]
Yes, but the model is pretty resource intensive actually right now. Hope further work on this will resolve that as well :smile:

**Comment by aniket-agarwal1999** [comment]
Everything seems good now, I'll merge it up.

**Comment by praeclarumjj3** [comment]
> Yes, but the model is pretty resource intensive actually right now. Hope further work on this will resolve that as well

Yes, it was mentioned as a challenge by the authors they plan to work on in the near future.

**Pull request merged**